cirq/protocols/decompose.py

```diff
@@ -0,0 +1,234 @@
+    TypeVar, List, Optional
+
+from typing_extensions import Protocol
+
+
+from cirq.type_workarounds import NotImplementedType
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    import cirq
+
+
+TDefault = TypeVar('TDefault')
+
+TError = TypeVar('TError', Exception)
+
+RaiseTypeErrorIfNotProvided = ([],)
+
+
+def _value_error_describing_bad_operation(op: 'cirq.Operation') -> ValueError:
+    return ValueError(
+        "Operation doesn't satisfy the given `keep` "
+        "but can't be decomposed: {!r}".format(op))
+
+
+class SupportsDecompose(Protocol):
+    """An object that may be describable by a unitary matrix."""
+
+    def _decompose_(self) -> Union['cirq.OP_TREE', NotImplementedType]:
+        return NotImplemented
+
+
+class SupportsDecomposeWithQubits(Protocol):
+    """A gate that may be describable by a unitary matrix."""
+
+    def _decompose_(self, qubits: Tuple['cirq.QubitId', ...]
+                    ) -> Union['cirq.OP_TREE', NotImplementedType]:
+        return NotImplemented
+
+
+def _default_decomposer(op: 'cirq.Operation'
+                        ) -> Union['cirq.OP_TREE', NotImplementedType]:
+    method = getattr(op, '_decompose_', None)
+    return NotImplemented if op is None else method()
+
+
+def _intercept_decompose(
+        op: 'cirq.Operation',
+        decomposers: List[Callable[['cirq.Operation'], 'cirq.OP_TREE']]
+) -> 'cirq.OP_TREE':
+    for d in decomposers:
+        r = d(op)
+        if r is not NotImplemented:
+            return r
+    return NotImplemented
+
+
+def decompose(
+    val: Any,
+    *,
+    intercepting_decomposer: Callable[['cirq.Operation'],
+                                      'cirq.OP_TREE'] = None,
+    decomposer: Callable[['cirq.Operation'],
+                         'cirq.OP_TREE'] = _default_decomposer,
+    keep: Callable[['cirq.Operation'], bool] = None,
+    on_stuck_raise: Optional[Union[
+        TError,
+        Callable[['cirq.Operation'], TError]]
+    ] = _value_error_describing_bad_operation
+) -> List['cirq.Operation']:
+    """Recursively decomposes a value into `cirq.Operation`s.
+
+    Args:
+        val: The value to decompose into operations.
+        intercepting_decomposer: An optional method that is called before the
+            main decomposer when doing intermediate decompositions. If
+            `intercepting_decomposer` isn't specified, or returns
+            `NotImplemented`, the decomposition falls back to the main
+            `decomposer`. Otherwise the result of `intercepting_decomposer`
+            takes priority and `decomposer` isn't even called.
+        decomposer: Decomposes operations into more basic operations, or returns
+            `NotImplemented` to indicate that an operation cannot be decomposed
+            anymore. Defaults to a decomposer that calls operations'
+            `_decompose_` method.
+        keep: A predicate that determines if the initial operation or
```

Comment:
keep feels a bit ambiguous to me.  It makes sense in terms of creating the list of outputs, but I'm not sure stoping decomposing would be thought of as "keep"ing.  Maybe
is_decomposed?
is_fully_decomposed?
or invert it: 
needs_more_decomposition?

Reply:
Acknowledged. I think this is a good name in the same sense that `key` is a good argument for `sorted`. It's a vague word that only works because the context pushes you toward one particular interpretation. It makes up for that by being short and memorable.

Reply:
OK let's watch if this confuses people.  I think the problem is that keep is close to "keep going" in my head.